File: java/leet_code/1219.java
/**
 * This class provides a solution for finding the maximum amount of gold that can be collected in a grid.
 * The grid is a 2D array where each cell represents a piece of gold. The goal is to find the path that collects the most gold.
 * The path can start and end at any cell in the grid, and can move up, down, left, or right from the current cell.
 * Each cell can only be visited once in a path.
 * <p>
 * Copyright © 2010-2024 <a href="https://github.com/jerodg/">JerodG</a>
 * <p>
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
 * or (at your option) any later version.
 * <p>
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
 * for more details.
 * <p>
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software. You should have received a copy of the SSPL along with this
 * program. If not, see <<a href="https://www.mongodb.com/licensing/server-side-public-license">SSPL</a>>.
 */
class Solution {
    /**
     * The maximum amount of gold collected so far.
     */
    public int max = 0;

    /**
     * Recursive method to find all possible paths and update the maximum gold collected.
     *
     * @param grid The 2D grid of gold pieces.
     * @param r    The current row in the grid.
     * @param c    The current column in the grid.
     * @param gold The amount of gold collected so far.
     */
    private void ways(final int[][] grid, final int r, final int c, final int gold) {
        // Update max if current gold is more
        if (gold > this.max) {
            this.max = gold;
        }
        // Return if current cell is empty
        if (grid[r][c] == 0) {
            return;
        }
        // Collect gold from current cell and mark it as empty
        final int a = grid[r][c];
        grid[r][c] = 0;
        // Explore all possible directions
        if (r - 1 >= 0) {
            this.ways(grid, r - 1, c, gold + a);
        }
        if (r + 1 <= grid.length - 1) {
            this.ways(grid, r + 1, c, gold + a);
        }
        if (c + 1 <= grid[0].length - 1) {
            this.ways(grid, r, c + 1, gold + a);
        }
        if (c - 1 >= 0) {
            this.ways(grid, r, c - 1, gold + a);
        }
        // Restore the gold in the current cell after exploring all paths
        grid[r][c] = a;
    }

    /**
     * Method to check if the grid contains any zero cells.
     * If there are no zero cells, it returns the total amount of gold in the grid.
     * Otherwise, it returns -1.
     *
     * @param grid The 2D grid of gold pieces.
     *
     * @return The total amount of gold if there are no zero cells, otherwise -1.
     */
    private static int gridWithNoZeros(final int[][] grid) {
        int count = 0;
        for (final int[] ints : grid) {
            for (int j = 0; j < grid[0].length; j++) {
                if (ints[j] == 0) {
                    return -1;
                } else count += ints[j];
            }
        }
        return count;
    }

    /**
     * Method to find the maximum amount of gold that can be collected in the grid.
     *
     * @param grid The 2D grid of gold pieces.
     *
     * @return The maximum amount of gold that can be collected.
     */
    public final int getMaximumGold(final int[][] grid) {
        // Check if the grid contains any zero cells
        final int count = Solution.gridWithNoZeros(grid);
        // If there are no zero cells, return the total amount of gold
        if (count != -1) return count;
        // Otherwise, find the maximum gold that can be collected
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                this.ways(grid, i, j, 0);
            }
        }
        return this.max;
    }
}

File: javascript/leet_code/1219.mjs
/**
 * Copyright ©2010-2024 <a href="https://github.com/jerodg/">JerodG</a>
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
 * for more details.
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software. You should have received a copy of the SSPL along with this
 * program. If not, see SSPL.
 */

/**
 * Package leet_code
 *
 * This file `1219.mjs` contains a function `getMaximumGold` that calculates the maximum amount of gold that can be
 * collected from a grid. The grid is represented as a 2D array of numbers, where each number represents the amount of
 * gold in a cell. The function uses depth-first search (DFS) to find the path that collects the maximum amount of
 * gold.
 */

/**
 * Function to calculate the maximum amount of gold that can be collected from a grid.
 *
 * @param {number[][]} grid - The grid represented as a 2D array of numbers.
 * @returns {number} - The maximum amount of gold that can be collected.
 */
const getMaximumGold = function (grid) {
    // Check if the grid contains a cell with a value of 0.
    const sum = isZeroPresent(grid);
    if (sum !== -1) {
        return sum;
    }

    // Initialize the maximum amount of gold that can be collected.
    let max = 0;

    // Iterate over each cell in the grid.
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            // Update the maximum amount of gold that can be collected.
            max = Math.max(max, maxPath(grid, i, j));
        }
    }

    // Return the maximum amount of gold that can be collected.
    return max;
};

/**
 * Function to check if the grid contains a cell with a value of 0.
 *
 * @param {number[][]} grid - The grid represented as a 2D array of numbers.
 * @returns {number} - The sum of the values in the grid if it does not contain a cell with a value of 0, -1 otherwise.
 */
const isZeroPresent = (grid) => {
    // Initialize the sum of the values in the grid.
    let sum = 0;

    // Iterate over each cell in the grid.
    for (const row of grid) {
        for (const val of row) {
            // If the cell has a value of 0, return -1.
            if (!val) {
                return -1;
            }

            // Update the sum of the values in the grid.
            sum += val;
        }
    }

    // Return the sum of the values in the grid.
    return sum;
};

/**
 * Function to calculate the maximum amount of gold that can be collected from a cell using DFS.
 *
 * @param {number[][]} grid - The grid represented as a 2D array of numbers.
 * @param {number} i - The row index of the cell.
 * @param {number} j - The column index of the cell.
 * @returns {number} - The maximum amount of gold that can be collected from the cell.
 */
const maxPath = (grid, i, j) => {
    // If the cell is outside the grid or has a value of 0, return 0.
    if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || !grid[i][j]) {
        return 0;
    }

    // Store the value of the cell and set it to 0 to mark it as visited.
    const val = grid[i][j];
    grid[i][j] = 0;

    // Calculate the maximum amount of gold that can be collected from the adjacent cells.
    const upPath = maxPath(grid, i, j - 1);
    const downPath = maxPath(grid, i, j + 1);
    const leftPath = maxPath(grid, i - 1, j);
    const rightPath = maxPath(grid, i + 1, j);

    // Restore the value of the cell.
    grid[i][j] = val;

    // Return the maximum amount of gold that can be collected from the cell.
    return val + Math.max(Math.max(upPath, downPath), Math.max(leftPath, rightPath));
};
